src/v/ssx/sharded_ptr.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "base/vassert.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sharded.hh>
+
+#include <memory>
+
+namespace ssx {
+
+/// A pointer that is safe to share between shards.
+///
+/// The pointer can be reset only by the home shard; other shards shall not
+/// observe the change until the update has run on their reactor.
+///
+/// As such, it is safe to maintain a pointer or reference to the pointee an any
+/// shard until the next yield point.
+template<typename T>
+class sharded_ptr {
+public:
+    sharded_ptr()
+      : _shard{ss::this_shard_id()} {}
+    ~sharded_ptr() {
+        if (*this) {
+            std::default_delete<T>{}(_state.local());
+        }
+    }
+
+    sharded_ptr(sharded_ptr&& other) noexcept = delete;
+    sharded_ptr(sharded_ptr const&) noexcept = delete;
+    sharded_ptr& operator=(sharded_ptr const&) = delete;
+    sharded_ptr& operator=(sharded_ptr&&) = delete;
+
+    T& operator*() const { return *_state.local(); }
+    T* operator->() const { return _state.local(); }
+    explicit operator bool() const {
+        return _state.local_is_initialized() && _state.local() != nullptr;
+    }
+
+    ss::future<> reset(std::unique_ptr<T> u = nullptr) {
+        assert_shard();
+        auto gh{_gate.hold()};
+        auto up{u.get()};
+
+        if (!_state.local_is_initialized()) {
+            if (u == nullptr) {
+                co_return;
+            }
+            co_return co_await _state.start(up).finally(
+              [this, u{std::move(u)}]() mutable noexcept {
+                  if (u.get() == _state.local()) {
+                      void(u.release());
+                  }
+              });
+        }
```

Comment:
Overall I think this method has several races, but I guess it depends on the exact semantics of the sharded_ptr. 

E.g., I would expect that when we are not doing any further `reset()` calls on the pointer, all shards eventually converge on the same value for the pointer, right? I don't think that's going to be the case now since when the owning shard does some concurrent resets, different shards may get a different final value depending on what the last writer was on that shard. This re-ordering could happen on the owning shard, if it suspends part way through `start()` or `invoke_on_others`, in the shard<->shard queues (if reordering is allowed there) or on the other shards since task order is guaranteed.

I think there are other races along these lines (all related to concurrent resets).

You could do something fancy with generation counter and ignore old updates, but a big fat mutex around reset should solve it? I think the mutex could also replace the gate: just take the mutex also in `stop`.

Reply:
Yeah, mutex works for me:
```c++
    ss::future<> reset(std::unique_ptr<T> u = nullptr) {
        assert_shard();
        auto mu{co_await _mutex.get_units()};
        auto up{u.get()};

        std::unique_ptr<T> l{
          !_state.local_is_initialized() ? nullptr : _state.local()};
        co_await (
          !_state.local_is_initialized()
            ? _state.start(up)
            : _state.invoke_on_all([up](T*& state) noexcept { state = up; }))
          .finally([this, l{std::move(l)}, u{std::move(u)}]() mutable noexcept {
              if (_state.local() == u.get()) {
                  void(u.release());
              } else if (_state.local() == l.get()) {
                  void(l.release());
              }
          });
    }

    ss::future<> stop() {
        co_await reset();
        co_await _mutex.with([this] { _mutex.broken(); });
        co_await _state.stop();
    }
```